### PR TITLE
Feature/xaptc 37

### PR DESCRIPTION
**Pull request by frankquinn**
Test Suite POST

### CODE REVIEW

File: uiServices/src/testSuiteServices.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+func TestSuiteCtx(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		next.ServeHTTP(w, r)
+	})
+
+}
+
+func postTestSuites(rw http.ResponseWriter, req *http.Request) {
+	configPathDir := GetConfigHeader(req)
```
Comment (jenny-flynn) [inline uiServices/src/testSuiteServices.go]: GetTestSuiteHeader? Different values right?

File: uiServices/src/testSuiteServices.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	}
+
+	if !FilePathExist(configPathDir) {
+		logrus.Error("File path does not exist", err)
+		rw.WriteHeader(http.StatusBadRequest)
+		return
+
+	}
+
+	if FilePathExist(configPathDir + testSuite.Name + ".xml") {
```
Comment (jenny-flynn) [inline uiServices/src/testSuiteServices.go]: I would replace the string contatenation with `fmt.Sprintf("%s%s.xml",configPathDir, testSuite.Name)`

File: uiServices/src/testSuiteServices.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+	goPath := os.Getenv("GOPATH")
+	schemaLoader := gojsonschema.NewReferenceLoader("file:///" + goPath + "/src/github.com/xtracdev/automated-perf-test/ui-src/src/assets/testSuite_schema.json")
+	documentLoader := gojsonschema.NewBytesLoader(testSuite)
+	logrus.Info(schemaLoader)
+	result, error := gojsonschema.Validate(schemaLoader, documentLoader)
+
+	if error != nil {
+		return false
+	}
+	if result.Valid() {
```
Comment (jenny-flynn) [inline uiServices/src/testSuiteServices.go]: remove this part. then at v end of function, just log it's valid and return true. no need for separate check.

File: uiServices/src/testSuiteServices_test.go [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+      "preThinkTime": 1000,
+      "postThinkTime": 2000,
+      "execWeight": "Infrequent"
+    },
+    {
+      "name":"file2.xml",
+      "preThinkTime": 1,
+      "postThinkTime": 10,
+      "execWeight": "Sparse"
+    }
```
Comment (jenny-flynn) [inline uiServices/src/testSuiteServices_test.go]: include a testcase with only name. this is valid too.

File: uiServices/src/testSuiteServices_test.go [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+	request, err := http.NewRequest(http.MethodPost, "/test-suites", reader)
+	request.Header.Set("configPathDir", filePath)
+
+	w := httptest.NewRecorder()
+	r.ServeHTTP(w, request)
+
+	if err != nil {
+		t.Error(err)
+
+		assert.Equal(t, w.Code, http.StatusCreated, "Error: Did Not Successfully Post")
```
Comment (jenny-flynn) [inline uiServices/src/testSuiteServices_test.go]: this should not be in the err check

File: uiServices/test/configFile.feature [hunk 6]
```diff
@@ -131,12 +131,12 @@ Feature: Create Configuration File
...
     And the header configsDirPath is "/uiServices/test/"
     And the file name is "GodogConfig.xml"
-    When I send a "GET" request to "/configs/GodogConfig"
+    When I send a "GET" request to "/configs/GodogConfig.xml"
     Then the response code should be 200
     And the response body should match json:
     """
       {
-       "apiName": "GodogConfig",
+       "apiName": "GodogConfig.xml",
```
Comment (jenny-flynn) [inline uiServices/test/configFile.feature]: no, this should not be ".xml" based on all config files i've seen so far

File: uiServices/test/configFile.feature [hunk 7]
```diff
@@ -170,11 +170,7 @@ Feature: Create Configuration File
     And the header configsDirPath is ""
     When I send a "GET" request to "/configs/GodogConfig"
     Then the response code should be 400
-
-
-  Scenario: Unsuccessful retrieval of config file (invalid URL)
```
Comment (jenny-flynn) [inline uiServices/test/configFile.feature]: why is this removed?

File: uiServices/test/configFile.feature [hunk 8]
```diff
@@ -360,7 +356,7 @@ Feature: Create Configuration File
     Given the config file "GodogConfig.xml" exists at "/uiServices/test/"
     Given the automated performance ui server is available
     And the header configsDirPath is "/uiServices/test/"
-    When I send "PUT" request to "/configs/GodogConfig" with body:
+    When I send "PUT" request to "/configs/GodogConfig.xml" with body:
```
Comment (jenny-flynn) [inline uiServices/test/configFile.feature]: not sure i'm happy .xml is being passed down. should we just pass down the name and assume xml? perhaps the server could assume xml, and we could pass down a header if this ever changed in the future.

File: uiServices/test/testSuite.feature [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+                                ###################################
+
+  Scenario: Successful creation of test Suite
+    Given there is no existing test file "GodogTestSuite.xml"
+    Given the automated performance ui server is available
+    And the header configsDirPath is "/uiServices/test/GodogTestSuite.xml"
+    When I send "POST" request to "/test-suites" with a body:
+         """
+      {
+  "name": "GodogTestSuite",
```
Comment (jenny-flynn) [inline uiServices/test/testSuite.feature]: fix formatting here.

File: uiServices/test/configFile.feature (hunk 8 above)
Comment (frankquinn) [reply uiServices/test/configFile.feature]: see email about this one!

File: uiServices/src/testSuiteServices.go [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+		return
+	}
+
+	if !validateTestSuiteJsonWithSchema(buf.Bytes()) {
+		rw.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	if !FilePathExist(testSuitePathDir) {
+		logrus.Error("File path does not exist", err)
```
Comment (jenny-flynn) [inline uiServices/src/testSuiteServices.go]: `logrus.Error("Directory path does not exist", err)`

File: uiServices/src/testSuiteServices.go [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+		rw.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	rw.WriteHeader(http.StatusCreated)
+}
+
+func validateTestSuiteJsonWithSchema(testSuite []byte) bool {
+	goPath := os.Getenv("GOPATH")
+	schemaLoader := gojsonschema.NewReferenceLoader("file:///" + goPath + "/src/github.com/xtracdev/automated-perf-test/ui-src/src/assets/testSuite_schema.json")
```
Comment (jenny-flynn) [inline uiServices/src/testSuiteServices.go]: we could abstract this out and pass in the appropriate schema and log message

File: uiServices/src/testSuiteServices_test.go [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+    },
+    {
+      "name":"file2.xml"
+    }
+  ]
+}
+`
+const TestSuiteMissingRequired = `
+{
+  "testCases": [
```
Comment (jenny-flynn) [inline uiServices/src/testSuiteServices_test.go]: add the name in so that `TestMissingRequiredTestSuitePost` is caught by` !validateTestSuiteJsonWithSchema(buf.Bytes()) in testSuiteServices.go`

File: uiServices/src/testSuiteServices_test.go [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+	request.Header.Set("testSuitePathDir", filePath)
+
+	w := httptest.NewRecorder()
+	r.ServeHTTP(w, request)
+
+	if err != nil {
+		t.Error(err)
+	}
+
+	assert.Equal(t, w.Code, http.StatusBadRequest, "Created, but should not have. File Should Alreday Exist")
```
Comment (jenny-flynn) [inline uiServices/src/testSuiteServices_test.go]: you should look at all these asserts and flip w.Code, http.StatusBadRequest so the (expected) and (actual) log correctly:

> 
> 	Error Trace:	testSuiteServices_test.go:78
> 	Error:		Not equal: 400 (expected)
> 			        != 201 (actual)
> 	Messages:	Error: Did Not Successfully Post

File: uiServices/test/testSuite.feature [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+    Given the automated performance ui server is available
+    And the header "testSuitePathDir" is "/uiServices/test/GodogTestSuite.xml"
+    When I send "POST" request to "/test-suites" with a body:
+     """
+      {
+       "name": "",
+       "testStrategy": "SuiteBased",
+       "testCases":[
+         {
+         "name":"file1.xml",
```
Comment (jenny-flynn) [inline uiServices/test/testSuite.feature]: we should go with config approach of appending .xml separately to the name, right? name and file name can be independent.

File: uiServices/src/testSuiteServices.go [hunk 15]
```diff
@@ -83,12 +83,12 @@ func validateTestSuiteJsonWithSchema(testSuite []byte) bool {
...
 	if !result.Valid() {
-		logrus.Error("**** The TestSuite document is not valid. see errors :")
+		logrus.Error("**** The "+structType+" document is not valid. see errors :")
 		for _, desc := range result.Errors() {
 			logrus.Error("- ", desc)
 			return false
 		}
 	}
-	logrus.Info("**** The TestSuite document is valid *****")
+	logrus.Info("**** The "+structType+" document is valid *****")
```
Comment (jenny-flynn) [inline uiServices/src/testSuiteServices.go]: logrus.Infof might be a better choice here

File: uiServices/src/testSuiteServices.go [hunk 16]
```diff
@@ -44,13 +44,13 @@ func postTestSuites(rw http.ResponseWriter, req *http.Request) {
 		return
 	}
 
-	if !validateTestSuiteJsonWithSchema(buf.Bytes()) {
+	if !ValidateJsonWithSchema(buf.Bytes(), "testSuite_schema.json", "TestSuite") {
```
Comment (jenny-flynn) [inline uiServices/src/testSuiteServices.go]: these should be constants

File: uiServices/src/configServices.go [hunk 17]
```diff
@@ -188,7 +162,7 @@ func putConfigs(rw http.ResponseWriter, req *http.Request) {
 	buf := new(bytes.Buffer)
 	buf.ReadFrom(req.Body)
 
-	if !validateJsonWithSchema(buf.Bytes()) {
+	if !ValidateJsonWithSchema(buf.Bytes(), "schema.json","Configurations") {
```
Comment (jenny-flynn) [inline uiServices/src/configServices.go]: constants

File: ui-src/src/assets/testSuite_schema.json [hunk 18]
```diff
@@ -10,12 +10,15 @@
...
+    "description": {
+      "type": "string"
+    },
     "testCases": {
       "type": "array",
       "items": {
         "type": "object",
         "additionalProperties": false,
-        "required": ["name"],
+        "required": ["name", "description"],
```
Comment (jenny-flynn) [inline ui-src/src/assets/testSuite_schema.json]: don't assume descriptions are required - maybe some don't have descriptions?

File: ui-src/src/assets/testSuite_schema.json [hunk 19]
```diff
@@ -1,6 +1,6 @@
 {
   "type": "object",
-  "required": ["name", "testStrategy"],
+  "required": ["name", "testStrategy", "description"],
```
Comment (jenny-flynn) [inline ui-src/src/assets/testSuite_schema.json]: don't assume descriptions are required - maybe some don't have descriptions?

**Comment by jenny-flynn** [comment]
Think all of ye should be on each other's PRs too. @frankquinn @angerooney11 @ColmCooney

File: uiServices/src/homepage.go [hunk 20]
```diff
@@ -58,6 +58,7 @@ func GetIndexPage() *chi.Mux {
 	})
 
 	router.Mount("/configs", routeConfigs())
+	router.Mount("/test-suites", routeTestSuites())
```
Comment (jenny-flynn) [inline uiServices/src/homepage.go]: great yep!

File: uiServices/src/configServices_test.go [hunk 21]
```diff
@@ -420,7 +420,7 @@ func TestInvalidJsonPut(t *testing.T) {
 		t.Error(err)
 	}
 
-	assert.Equal(t, w.Code, http.StatusBadRequest, "Sucessfully updated. Field data type should have been incorrect so update should occur")
+	assert.Equal(t, http.StatusBadRequest, w.Code, "Sucessfully updated. Field data type should have been incorrect so update should occur")
```
Comment (jenny-flynn) [inline uiServices/src/configServices_test.go]: shouldn't?

**Comment by frankquinn** [comment]
All requested changes have been implemented
